File: takengine/takengine/src/main/cpp/raster/mosaic/MultiplexingMosaicDatabaseCursor2.cpp
#include "raster/mosaic/MultiplexingMosaicDatabaseCursor2.h"

#include <cmath>
#include <cstring>

using namespace TAK::Engine::Raster::Mosaic;

using namespace TAK::Engine::Core;
using namespace TAK::Engine::Util;

MultiplexingMosaicDatabaseCursor2::MultiplexingMosaicDatabaseCursor2() NOTHROWS :
impl(new Impl())
{}

TAKErr MultiplexingMosaicDatabaseCursor2::add(MosaicDatabase2::CursorPtr &&cursor) NOTHROWS
{
    TAKErr code(TE_Ok);
    this->impl->invalid.insert(cursor.get());
    this->impl->cursors.push_back(std::move(cursor));
    return code;
}

TAKErr MultiplexingMosaicDatabaseCursor2::moveToNext() NOTHROWS
{
    return this->impl->moveToNext();
}

TAKErr MultiplexingMosaicDatabaseCursor2::getUpperLeft(GeoPoint2 *value) NOTHROWS
{
    return this->impl->current->getUpperLeft(value);
}

TAKErr MultiplexingMosaicDatabaseCursor2::getUpperRight(GeoPoint2 *value) NOTHROWS
{
    return this->impl->current->getUpperRight(value);
}

TAKErr MultiplexingMosaicDatabaseCursor2::getLowerRight(GeoPoint2 *value) NOTHROWS
{
    return this->impl->current->getLowerRight(value);
}

TAKErr MultiplexingMosaicDatabaseCursor2::getLowerLeft(GeoPoint2 *value) NOTHROWS
{
    return this->impl->current->getLowerLeft(value);
}

TAKErr MultiplexingMosaicDatabaseCursor2::getMinLat(double *value) NOTHROWS
{
    return this->impl->current->getMinLat(value);
}

TAKErr MultiplexingMosaicDatabaseCursor2::getMinLon(double *value) NOTHROWS
{
    return this->impl->current->getMinLon(value);
}

TAKErr MultiplexingMosaicDatabaseCursor2::getMaxLat(double *value) NOTHROWS
{
    return this->impl->current->getMaxLat(value);
}

TAKErr MultiplexingMosaicDatabaseCursor2::getMaxLon(double *value) NOTHROWS
{
    return this->impl->current->getMaxLon(value);
}

TAKErr MultiplexingMosaicDatabaseCursor2::getPath(const char **value) NOTHROWS
{
    return this->impl->current->getPath(value);
}

TAKErr MultiplexingMosaicDatabaseCursor2::getType(const char **value) NOTHROWS
{
    return this->impl->current->getType(value);
}

TAKErr MultiplexingMosaicDatabaseCursor2::getMinGSD(double *value) NOTHROWS
{
    return this->impl->current->getMinGSD(value);
}

TAKErr MultiplexingMosaicDatabaseCursor2::getMaxGSD(double *value) NOTHROWS
{
    return this->impl->current->getMaxGSD(value);
}

TAKErr MultiplexingMosaicDatabaseCursor2::getWidth(int *value) NOTHROWS
{
    return this->impl->current->getWidth(value);
}

TAKErr MultiplexingMosaicDatabaseCursor2::getHeight(int *value) NOTHROWS
{
    return this->impl->current->getHeight(value);
}

TAKErr MultiplexingMosaicDatabaseCursor2::getId(int *value) NOTHROWS
{
    // XXX -
    return this->impl->current->getId(value);
}

TAKErr MultiplexingMosaicDatabaseCursor2::getSrid(int *value) NOTHROWS
{
    return this->impl->current->getSrid(value);
}

TAKErr MultiplexingMosaicDatabaseCursor2::isPrecisionImagery(bool *value) NOTHROWS
{
    return this->impl->current->isPrecisionImagery(value);
}

/**************************************************************************/

MultiplexingMosaicDatabaseCursor2::Impl::Impl() NOTHROWS :
current(nullptr)
{}

TAKErr MultiplexingMosaicDatabaseCursor2::Impl::moveToNext() NOTHROWS
{
    TAKErr code(TE_Ok);
    code = TE_Ok;
    // update entries for any cursors marked 'invalid'
    std::set<MosaicDatabase2::Cursor *>::iterator cursor;
    for (cursor = this->invalid.begin(); cursor != this->invalid.end(); cursor++) {
        code = (*cursor)->moveToNext();
        if (code == TE_Ok) {
            this->pendingResults.insert(Entry(*cursor));
        }
        else if (code == TE_Done) {
            // the cursor has been exhausted, drop it after we exit the loop
            code = TE_Ok;
        }
        TE_CHECKBREAK_CODE(code);
    }
    TE_CHECKRETURN_CODE(code);

    // all cursors are now valid
    this->invalid.clear();

    // if there are no pending results, we've exhausted the input
    if (this->pendingResults.empty())
        return TE_Done;

    // remove the cursor pointing at the frame with the highest
    // resolution (numerically lowest GSD value)
    Entry entry = *this->pendingResults.begin();
    this->pendingResults.erase(this->pendingResults.begin());
    // now removed, we want to re-evaluate the next call to 'moveToNext'
    // so place it in the invalid list
    this->invalid.insert(entry.cursor);
    // reset our current pointer
    this->current = entry.cursor;

    return code;
}

MultiplexingMosaicDatabaseCursor2::Entry::Entry(MosaicDatabase2::Cursor *cursor_) NOTHROWS :
    gsd(NAN),
    path(nullptr),
    cursor(cursor_)
{
    TAKErr code(TE_Ok);

    this->cursor->getMaxGSD(&this->gsd);

    const char *cpath;
    code = this->cursor->getPath(&cpath);

    this->path = cpath;
}

bool MultiplexingMosaicDatabaseCursor2::Entry::operator<(const Entry &other) const
{
    if (this->gsd < other.gsd)
        return true;
    else if (this->gsd > other.gsd)
        return false;
    else // this.gsd == other.gsd
         // XXX - should consider something faster than string comparison
        return (strcmp(this->path, other.path) < 0);
}

File: takkernel/engine/src/main/java/com/atakmap/map/layer/raster/mosaic/MultiplexingMosaicDatabaseCursor2.java
package com.atakmap.map.layer.raster.mosaic;

import java.util.Collection;
import java.util.Comparator;
import java.util.HashMap;
import java.util.LinkedList;
import java.util.Map;
import java.util.SortedSet;
import java.util.TreeSet;

import com.atakmap.coremap.maps.coords.GeoPoint;
import com.atakmap.database.RowIterator;
import com.atakmap.database.RowIteratorWrapper;

import com.atakmap.coremap.log.Log;

public final class MultiplexingMosaicDatabaseCursor2 extends RowIteratorWrapper implements MosaicDatabase2.Cursor {

    static public final String TAG = "MultiplexingMosaicDatabaseCursor2";

    private final static Map<MosaicDatabase2.QueryParameters.Order, Comparator<Entry>> ORDER_COMPARATORS = new HashMap<>();
    static {
        ORDER_COMPARATORS.put(MosaicDatabase2.QueryParameters.Order.MaxGsdDesc, new Comparator<Entry>() {
            @Override
            public int compare(Entry lhs, Entry rhs) {
                if(lhs.maxGsd < rhs.maxGsd)
                    return -1;
                else if(lhs.maxGsd > rhs.maxGsd)
                    return 1;
                // XXX - should consider something faster than string comparison
                return lhs.path.compareTo(rhs.path);
            }
        });
        ORDER_COMPARATORS.put(MosaicDatabase2.QueryParameters.Order.MaxGsdAsc, new Comparator<Entry>() {
            @Override
            public int compare(Entry lhs, Entry rhs) {
                if(lhs.maxGsd < rhs.maxGsd)
                    return 1;
                else if(lhs.maxGsd > rhs.maxGsd)
                    return -1;
                // XXX - should consider something faster than string comparison
                return lhs.path.compareTo(rhs.path);
            }
        });
        ORDER_COMPARATORS.put(MosaicDatabase2.QueryParameters.Order.MinGsdDesc, new Comparator<Entry>() {
            @Override
            public int compare(Entry lhs, Entry rhs) {
                if(lhs.minGsd < rhs.minGsd)
                    return -1;
                else if(lhs.minGsd > rhs.minGsd)
                    return 1;
                // XXX - should consider something faster than string comparison
                return lhs.path.compareTo(rhs.path);
            }
        });
        ORDER_COMPARATORS.put(MosaicDatabase2.QueryParameters.Order.MinGsdAsc, new Comparator<Entry>() {
            @Override
            public int compare(Entry lhs, Entry rhs) {
                if(lhs.minGsd < rhs.minGsd)
                    return 1;
                else if(lhs.minGsd > rhs.minGsd)
                    return -1;
                // XXX - should consider something faster than string comparison
                return lhs.path.compareTo(rhs.path);
            }
        });
    }

    private final Impl impl;

    public MultiplexingMosaicDatabaseCursor2(Collection<MosaicDatabase2.Cursor> cursors, MosaicDatabase2.QueryParameters.Order order) {
        super(new Impl(cursors, order));
        
        this.impl = (Impl)this.filter;
    }

    @Override
    public GeoPoint getUpperLeft() {
        return this.impl.current.getUpperLeft();
    }

    @Override
    public GeoPoint getUpperRight() {
        return this.impl.current.getUpperRight();
    }

    @Override
    public GeoPoint getLowerRight() {
        return this.impl.current.getLowerRight();
    }

    @Override
    public GeoPoint getLowerLeft() {
        return this.impl.current.getLowerLeft();
    }

    @Override
    public double getMinLat() {
        return this.impl.current.getMinLat();
    }

    @Override
    public double getMinLon() {
        return this.impl.current.getMinLon();
    }

    @Override
    public double getMaxLat() {
        return this.impl.current.getMaxLat();
    }

    @Override
    public double getMaxLon() {
        return this.impl.current.getMaxLon();
    }

    @Override
    public String getPath() {
        return this.impl.current.getPath();
    }

    @Override
    public String getType() {
        return this.impl.current.getType();
    }

    @Override
    public double getMinGSD() {
        return this.impl.current.getMinGSD();
    }

    @Override
    public double getMaxGSD() {
        return this.impl.current.getMaxGSD();
    }

    @Override
    public int getWidth() {
        return this.impl.current.getWidth();
    }

    @Override
    public int getHeight() {
        return this.impl.current.getHeight();
    }

    @Override
    public int getId() {
        // XXX - 
        return this.impl.current.getId();
    }
    
    @Override
    public MosaicDatabase2.Frame asFrame() {
        return this.impl.current.asFrame();
    }
    
    @Override
    public int getSrid() {
        return this.impl.current.getSrid();
    }
    
    @Override
    public boolean isPrecisionImagery() {
        return this.impl.current.isPrecisionImagery();
    }

    /**************************************************************************/
    
    private static class Impl implements RowIterator {
        private final Collection<MosaicDatabase2.Cursor> cursors;
        public MosaicDatabase2.Cursor current;

        private SortedSet<Entry> pendingResults;
        private Collection<MosaicDatabase2.Cursor> invalid;

        public Impl(Collection<MosaicDatabase2.Cursor> cursors, MosaicDatabase2.QueryParameters.Order order) {
            this.cursors = cursors;
            this.current = null;
            if(order == null)
                order = MosaicDatabase2.QueryParameters.Order.MaxGsdDesc;
            this.pendingResults = new TreeSet<Entry>(ORDER_COMPARATORS.get(order));
            this.invalid = new LinkedList<MosaicDatabase2.Cursor>(this.cursors);
        }

        @Override
        public boolean moveToNext() {
            // update entries for any cursors marked 'invalid'
            for(MosaicDatabase2.Cursor cursor : this.invalid) {
                try { 
                    if(cursor.moveToNext())
                        this.pendingResults.add(new Entry(cursor));
                } catch (Exception e) { 
                    Log.e(TAG, "bad cursor encountered, will be cleared", e);
                }
            }
            // all cursors are now valid
            this.invalid.clear();
            
            // if there are no pending results, we've exhausted the input
            if(this.pendingResults.isEmpty())
                return false;
            
            // remove the cursor pointing at the frame with the highest
            // resolution (numerically lowest GSD value)
            Entry entry =  this.pendingResults.first();
            this.pendingResults.remove(entry);
            // now removed, we want to re-evaluate the next call to 'moveToNext'
            // so place it in the invalid list
            this.invalid.add(entry.cursor);
            // reset our current pointer
            this.current = entry.cursor;
            
            return true;
        }

        @Override
        public void close() {
            for(MosaicDatabase2.Cursor cursor : this.cursors)
                cursor.close();
            this.cursors.clear();
            
            this.invalid.clear();
            this.pendingResults.clear();
        }
        @Override
        public boolean isClosed() {
            return this.cursors.isEmpty();
        }
    }

    private final static class Entry {
        final double minGsd;
        final double maxGsd;
        final String path;
        final MosaicDatabase2.Cursor cursor;
        
        public Entry(MosaicDatabase2.Cursor cursor) {
            this.minGsd = cursor.getMinGSD();
            this.maxGsd = cursor.getMaxGSD();
            this.path = cursor.getPath();
            this.cursor = cursor;
        }

        @Override
        public boolean equals(Object o) {
            if (this == o) return true;
            if (o == null || getClass() != o.getClass()) return false;

            Entry entry = (Entry) o;

            if (Double.compare(entry.maxGsd, maxGsd) != 0) return false;
            if (path != null ? !path.equals(entry.path) : entry.path != null) return false;
            return cursor != null ? cursor.equals(entry.cursor) : entry.cursor == null;
        }

        @Override
        public int hashCode() {
            int result;
            long temp;
            temp = Double.doubleToLongBits(maxGsd);
            result = (int) (temp ^ (temp >>> 32));
            result = 31 * result + (path != null ? path.hashCode() : 0);
            result = 31 * result + (cursor != null ? cursor.hashCode() : 0);
            return result;
        }
    }
}
